File: java/project/core-java/src/main/java/org/tby/fourdk/core/query/bus/QueryLoggingMiddleware.java
package org.tby.fourdk.core.query.bus;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tby.fourdk.core.query.Query;
import org.tby.fourdk.core.query.Response;

public class QueryLoggingMiddleware implements QueryBus{
    private static final Logger LOGGER = LoggerFactory.getLogger(QueryLoggingMiddleware.class);

    private QueryBus queryBus;

    public QueryLoggingMiddleware(QueryBus queryBus) {
        this.queryBus = queryBus;
    }

    @Override
    public Response dispatch(Query query) {
        var begin =  System.currentTimeMillis();

        var type = query.getClass().getSimpleName();
        LOGGER.info("Handling query of type [{}] [{}]", type, query);
        Response queryResponse;
        try {
            queryResponse = this.queryBus.dispatch(query);
        } catch (RuntimeException e) {
            LOGGER.info("An error occurred while handling query [{}] [{}]", type, query);
            throw e;
        }
        var end = System.currentTimeMillis();

        LOGGER.info("Query of type [{}] [{}] has been handled in [{}] ms.",
                type, query.toString(), queryHandlerProcessTime(begin, end));

        return queryResponse;
    }

    private long queryHandlerProcessTime(long begin, long end) {
        return end - begin;
    }
}

File: rust/project/core-rust/src/dddk/query/bus_impl/query_logging_middleware.rs
use log::{error, info};
use crate::dddk::aliases::ResponseFromHandler;
use crate::dddk::query::query::Query;
use crate::dddk::query::query_bus::QueryBus;

pub struct QueryLoggingMiddleware {
    query_bus: Box<dyn QueryBus>,
}

impl QueryLoggingMiddleware {
    pub fn new(query_bus: Box<dyn QueryBus>) -> QueryLoggingMiddleware {
        QueryLoggingMiddleware {
            query_bus
        }
    }
}

impl QueryBus for QueryLoggingMiddleware {
    fn dispatch<'b>(&self, query: &'b dyn Query) -> ResponseFromHandler {
        info!("Dispatching a query [{}] [{:?}].",
              query.get_query_name(),
              query);
        let response = self.query_bus.dispatch(query);
        if response.is_err() {
            let error = response.err().unwrap();
            let error_message = error.to_string();
            error!(
                "An error has occurred when dispatching query [{}] [{:?}]: {}",
                query.get_query_name(),
                query,
                error_message
            );
            return Err(error);
        }
        let response = response.unwrap();
        info!(
            "Query[{}] [{:?}] has been handled and has returned response [{}].",
            query.get_query_name(),
            query,
            response.get_response_name()
        );
        return Ok(response);
    }
}

